sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,216 @@
+        mut self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        let cmd = build_listen_all_query(channels);
+        let _ = self.send(cmd.as_str()).await?;
+        Ok(PgListener::new(self))
+    }
+}
+
+impl PgPool {
+    /// Fetch a new connection from the pool and register it as a listener on the specified channel.
+    pub async fn listen(&self, channel: impl AsRef<str>) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        let cmd = build_listen_all_query(&[channel]);
+        let _ = conn.send(cmd.as_str()).await?;
+        Ok(PgListener::new(conn))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        &self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        let cmd = build_listen_all_query(channels);
+        let _ = conn.send(cmd.as_str()).await?;
+        Ok(PgListener::new(conn))
+    }
+}
+
+impl PgPoolConnection {
+    /// Register this connection as a listener on the specified channel.
+    ///
+    /// If an error is returned here, the connection will be dropped.
+    pub async fn listen(mut self, channel: impl AsRef<str>) -> Result<PgListener<Self>> {
+        let cmd = build_listen_all_query(&[channel]);
+        let _ = self.send(cmd.as_str()).await?;
+        Ok(PgListener::new(self))
+    }
+
+    /// Register this connection as a listener on all of the specified channels.
+    ///
+    /// If an error is returned here, the connection will be dropped.
+    pub async fn listen_all(
+        mut self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        let cmd = build_listen_all_query(channels);
+        let _ = self.send(cmd.as_str()).await?;
+        Ok(PgListener::new(self))
+    }
+}
+
+/// A stream of async database notifications.
+///
+/// Notifications will always correspond to the channel(s) specified this object is created.
+pub struct PgListener<C>(C);
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(conn: C) -> Self {
+        Self(conn)
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: DerefMut<Target = PgConnection>,
+{
+    /// Get the next async notification from the database.
+    pub async fn next(&mut self) -> Result<NotifyMessage> {
+        loop {
+            match (&mut self.0).receive().await? {
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                Some(msg) => {
+                    return Err(protocol_err!(
+                        "unexpected message received from database {:?}",
+                        msg
+                    )
+                    .into())
+                }
+                None => continue,
```

Comment:
`None` from `receive` means the connection was closed I believe.

Reply:
I think this reveals an aspect of this design overall which makes me think that we should be treating this as an actual `Stream`. It could still be used in all of the same ways which have been suggested, but semantically this pattern is a stream pattern whether we use a `Stream` implementation or not.

If we use a stream implementation, then returning `None` is the logical ending point of the stream. Without using a stream, users may end up in a state of confusion about how to handle conditions where the underlying connection has been closed. Should this interface just continually return an error if `None` is hit here on `L111`? What if the user doesn't handle it properly?

With a stream impl, users can still do:
```rust
while let Some(notification) = notify_stream.next().await {
// ...
}
```
See https://docs.rs/futures/0.3.1/futures/stream/trait.StreamExt.html#method.next for more details.

We could even design the stream to be `Stream<Item=PgNotification>`; IE, no error condition. If an error comes up within the stream impl, we can just transform it into a `None` (as there should really be no actual error conditions in the listener). If running from a pool, we can retrieve a new connection and resume listening any time a connection is lost.

With the non-stream interface, we are exposing a stream-like interface with none of the standard expectations which folks will have when working with async Rust. The implementation of the stream interface should be pretty simple. I'm thinking we could use [stream::unfold](https://docs.rs/futures/0.3.1/futures/stream/fn.unfold.html) to wrap the calls to `receive`. The init state would just be `()` probably, as we don't really need a seed value. Using unfold just allows us to stay in async/await land instead of having to work with `poll_next` (which can be a PITA).

@mehcode @abonander thoughts?

Reply:
Ok, going to get this all wrapped up in the `PgListener.receive` interface (which the stream impl will be based upon as well). I have a TODO item for it at the top of the PR now as well.